### PR TITLE
BRIDGE-2322: on consent, tag user/enroll in substudy

### PR DESCRIPTION
**Pull request by alxdarksage**
And remove data groups on withdrawal. I also rewrote the ConsentServiceMockTest because it was trying to mock all the behaviors of the dependencies and had become impossible to work with.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulation.java [hunk 1]
```diff
@@ -173,10 +180,35 @@ public void setAutoSendConsentSuppressed(boolean autoSendConsentSuppressed) {
         this.autoSendConsentSuppressed = autoSendConsentSuppressed;
     }
     
+    /** {@inheritDoc */
+    public void setDataGroupsAssignedWhileConsented(Set<String> dataGroups) {
+        if (dataGroups != null) {
+            this.dataGroupsAssignedWhileConsented = new HashSet<>(dataGroups);    
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulation.java]: Is the correct behavior here to ignore nulls or to replace them with empty sets? (Similarly with the substudies set.)

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java [hunk 2]
```diff
@@ -102,9 +112,21 @@ private Subpopulation makeSubpopulation() {
...
     }
     
+    @Test
+    public void cannotNullifySets() {
+        Subpopulation subpop = new DynamoSubpopulation();
+        subpop.setDataGroupsAssignedWhileConsented(null);
+        subpop.setSubstudyIdsAssignedOnConsent(null);
+        
+        assertTrue(subpop.getDataGroupsAssignedWhileConsented().isEmpty());
+        assertTrue(subpop.getSubstudyIdsAssignedOnConsent().isEmpty());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java:132]: Is the correct behavior to ignore nulls or to turn them into empty sets? If the correct behavior is to ignore nulls, should we specify a non-empty set to verify that the null is actually ignored?

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 3]
```diff
@@ -93,555 +75,147 @@
...
-    public void consentUpdatesSession() throws Throwable {
-        // Note that even if there are two required consent that should be signed, the sharing scope is 
-        // set for this user and data would theoretically be exported for the user. However that same user 
-        // is still going to get 412s from other API endpoints because they are not consented, so they will
-        // not be able to send data to the server yet.
-        String json = createJson("{'name':'Jack Aubrey','birthdate':'1970-10-10','imageData':'data:asdf'"+
-                ",'imageMimeType':'image/png','scope':'sponsors_and_partners'}");
+    public void giveV3() throws Exception {
+        String json = createJson("{'name':'Jack Aubrey','birthdate':'1970-10-10',"+
+                "'imageData':'data:asdf','imageMimeType':'image/png','scope':'no_sharing'}");
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:192]: giveV1() and giveV2() need unit tests.

File: test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java [hunk 4]
```diff
@@ -319,6 +321,21 @@ public void withdrawConsentWithParticipant() throws Exception {
                 email.getMessageParts().get(0).getContent());
     }
     
+    @Test
+    public void withdrawConsentRemovesDataGroups() throws Exception {
+        account.setConsentSignatureHistory(SUBPOP_GUID, ImmutableList.of(CONSENT_SIGNATURE));
+        
+        when(subpopulation.getDataGroupsAssignedWhileConsented()).thenReturn(TestConstants.USER_DATA_GROUPS);
+        when(subpopService.getSubpopulation(study.getStudyIdentifier(), SUBPOP_GUID)).thenReturn(subpopulation);
+        when(accountDao.getAccount(any())).thenReturn(account);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java:345]: The account should start with the data groups in question (and maybe a few extra ones), so we can verify that they are, in fact, removed.

Similarly for withdrawFromStudyRemovesDataGroups().

File: test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java [hunk 5]
```diff
@@ -65,6 +68,8 @@ public void testValidation() {
             assertMessage(e, "noneOfGroups", " 'wrongGroup' is not in enumeration");
             assertMessage(e, "allOfSubstudyIds", " 'substudyC' is not in enumeration");
             assertMessage(e, "noneOfSubstudyIds", " 'substudyD' is not in enumeration");
+            assertMessage(e, "dataGroupsAssignedWhileConsented", " 'dataGroup3' is not in enumeration: group1, group2");
+            assertMessage(e, "substudyIdsAssignedOnConsent", " 'substudyC' is not in enumeration: substudyA, substudyB");
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java:75]: We also need tests for valid empty sets and valid non-empty sets.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 6]
```diff
@@ -93,555 +75,147 @@
...
-        Result result = controller.giveV2();
-        TestUtils.assertResult(result,  201);
-        
-        assertConsentInSession(result, SharingScope.NO_SHARING, DEFAULT_SUBPOP_GUID);
-
-        verify(consentService).consentToResearch(eq(study), eq(DEFAULT_SUBPOP_GUID), eq(participant),
-                signatureCaptor.capture(), eq(SharingScope.NO_SHARING), eq(true));
-        validateSignature(signatureCaptor.getValue());
+        verify(accountDao).editAccount(eq(TestConstants.TEST_STUDY), eq(HEALTH_CODE), any());
+        verify(sessionUpdateService).updateSharingScope(session, SharingScope.ALL_QUALIFIED_RESEARCHERS);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:142]: You should mock editAccount() to call through to the passed in function pointer and also verify that account.setSharingScope() is called.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 7]
```diff
@@ -93,555 +75,147 @@
...
-        
-        when(consentService.withdrawConsent(eq(study), eq(DEFAULT_SUBPOP_GUID), eq(participant), eq(context), any(),
-                anyLong())).thenAnswer(createAnswer(false, DEFAULT_SUBPOP_GUID));
-
-        Result result = controller.withdrawConsent();
-        TestUtils.assertResult(result,  200);
-        
-        assertWithdrawnInSession(result, SharingScope.NO_SHARING, DEFAULT_SUBPOP_GUID);
+    public void getConsentSignatureV2() throws Exception {
+        when(consentService.getConsentSignature(study, SUBPOP_GUID, USER_ID)).thenReturn(SIGNATURE);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:176]: We should also have a test for getConsentSignatureV1, to make sure the code is being exercised and passes in the correct guid.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 8]
```diff
@@ -93,555 +75,147 @@
...
+        verify(cacheProvider).setUserSession(updatedSession);
     }
     
     @Test
-    public void consentSignatureHasServerGeneratedSignedOnValue() throws Exception {
-        // This signedOn property should be ignored, it is always set on the server
-        String json = createJson("{'name':'Jack Aubrey','birthdate':'1970-10-10','signedOn':1000,"+
-                "'imageData':'data:asdf','imageMimeType':'image/png','scope':'no_sharing'}");
+    public void withdrawConsentV2() throws Exception {
+        TestUtils.mockPlay().withMockResponse().withBody(WITHDRAWAL).mock();
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:310]: withdrawConsentV1 needs unit test.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 9]
```diff
@@ -93,555 +75,147 @@
...
-        assertEquals("Jack Aubrey", node.get("name").asText());
-        assertEquals("1970-10-10", node.get("birthdate").asText());
-        assertEquals("ConsentSignature", node.get("type").asText());
-        assertEquals("data:asdf", node.get("imageData").asText());
-        assertEquals("image/png", node.get("imageMimeType").asText());
-        // no signedOn value when serializing
-    }
-    
-    @Test(expected = EntityNotFoundException.class)
-    public void givingConsentToInvalidSubpopulation() throws Exception {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:332]: We should keep this unit test.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 3 above)
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:192]: Should we add a signedOn to make sure the server is properly ignoring it?

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 10]
```diff
@@ -93,555 +75,147 @@
...
-        // withdrawing the optional consent does not turn off sharing 
-        Result result = controller.withdrawConsentV2(SUBPOP_GUID.getGuid());
-        
-        assertWithdrawnInSession(result, SharingScope.SPONSORS_AND_PARTNERS, SUBPOP_GUID);
-    }
-    
-    @Test
-    @SuppressWarnings("deprecation")
-    public void dataSharingSuspendedUpdatesSession() throws Exception {
-        TestUtils.mockEditAccount(accountDao, account);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java]: We should keep this unit test.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 11]
```diff
@@ -93,555 +75,147 @@
...
-        assertEquals("no_sharing", node.get("sharingScope").asText());
-        assertFalse(node.get("dataSharing").asBoolean());
-
-        verify(account).setSharingScope(SharingScope.NO_SHARING);
-    }
-    
-    @Test
-    @SuppressWarnings("deprecation")
-    public void dataSharingResumedUpdatesSession() throws Exception {
-        TestUtils.mockEditAccount(accountDao, account);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java]: We should keep this unit test.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulation.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulation.java]: null should empty the list... fixed.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java (hunk 2 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java:132]: Added

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 6 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:142]: What I found worked for this was to capture the lambda... it has a closure that includes the correct sharing scope, such that every time an account is passed to the closure, it should be edited in the desired way... and thus we can test the behavior that way.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 7 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:176]: Added

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 9 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:332]: Added back with an adjustment to the consent status map so it throws correctly.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 3 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:192]: Added

Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:192]: I added a spurious signedOn value in the V3 test (all the methods eventually call the same code) and it continues to match verification because the signature does not deserialize the value. (IOW: added).

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 8 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java:310]: Added

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 10 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java]: We can. Added mock implementation of the behavior of the sessionUpdatedService as well as some further verification.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 11 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java]: Added back with changes to work

File: test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java (hunk 4 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java:345]: Added data groups that are and are not in the list of assigned data groups while consented, verified only the list is removed.

File: test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java (hunk 5 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java:75]: Added

**Review by alxdarksage — COMMENTED** [review]
Updated based on CR

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java [hunk 12]
```diff
@@ -102,9 +112,24 @@ private Subpopulation makeSubpopulation() {
...
     }
     
+    @Test
+    public void cannotNullifySets() {
+        Subpopulation subpop = new DynamoSubpopulation();
+        // Set some values to verify that null resets these to the empty set
+        subpop.setDataGroupsAssignedWhileConsented(ImmutableSet.of("A"));
+        subpop.setSubstudyIdsAssignedOnConsent(ImmutableSet.of("B"));
+        subpop.setDataGroupsAssignedWhileConsented(null);
+        subpop.setSubstudyIdsAssignedOnConsent(null);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java:130]: In between the sets, we should verify that the collections have the expected data in them.

File: test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java [hunk 13]
```diff
@@ -407,6 +427,23 @@ public void withdrawFromStudyWithPhone() {
         verify(notificationsService).deleteAllRegistrations(study.getStudyIdentifier(), HEALTH_CODE);
     }
     
+    @Test
+    public void withdrawFromStudyRemovesDataGroups() {
+        account.setConsentSignatureHistory(SUBPOP_GUID, ImmutableList.of(CONSENT_SIGNATURE));
+
+        account.setDataGroups(new HashSet<>(TestConstants.USER_DATA_GROUPS));
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java]: Add additional data groups, so only the ones specified by the subpopulation are removed.

File: test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java [hunk 14]
```diff
@@ -65,9 +68,39 @@ public void testValidation() {
...
+        Subpopulation subpop = Subpopulation.create();
+        subpop.setStudyIdentifier("test-study");
+        subpop.setGuidString("AAA");
+        subpop.setName("Name");
+        subpop.setDataGroupsAssignedWhileConsented(ImmutableSet.of());
+        subpop.setSubstudyIdsAssignedOnConsent(ImmutableSet.of());
+        
+        Validate.entityThrowingException(validator, subpop);
+        assertTrue(subpop.getDataGroupsAssignedWhileConsented().isEmpty());
+        assertTrue(subpop.getSubstudyIdsAssignedOnConsent().isEmpty());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java:90]: We also need a test with valid non-empty lists.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java [hunk 15]
```diff
@@ -634,52 +376,35 @@ public void dataSharingSuspendedUpdatesSession() throws Exception {
...
 
         verify(account).setSharingScope(SharingScope.NO_SHARING);
+        
+        verify(accountDao).editAccount(eq(TestConstants.TEST_STUDY), eq(HEALTH_CODE),
+                accountConsumerCaptor.capture());
+        
+        Consumer<Account> accountConsumer = accountConsumerCaptor.getValue();
+        Account updatedAccount = Account.create();
+        accountConsumer.accept(updatedAccount);
+        assertEquals(SharingScope.NO_SHARING, updatedAccount.getSharingScope());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java]: This block doesn't actually do anything. The default sharing scope is no_sharing. This is also redundant with the block above, which already verifies that account.setSharingScope() was called.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java (hunk 12 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationTest.java:130]: Added

File: test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java (hunk 13 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ConsentServiceMockTest.java]: Added

File: test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java (hunk 14 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/validators/SubpopulationValidatorTest.java:90]: Added to the valid subpopulation test.

File: test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java (hunk 15 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/ConsentControllerMockedTest.java]: Sorry you asked about testing the lambda... I actually am not sure how to verify the lambda functionality beyond what the test originally did... verified that the setter on the account was called with the correct sharing status. Removed.

**Review by DwayneJengSage — APPROVED** [review]
